File: tetris.py
import numpy as np
import random


MOVE_LEFT = 0
MOVE_RIGHT = 1
ROTATE = 2
NOTHING = 3

MAX_ACTIONS = 10

TETRIMINOS = [
    np.array([
        # L Tetrimino
        [
            [1, 0, 0],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 1],
            [0, 1, 0],
            [0, 1, 0]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [0, 0, 1]
        ],
        [
            [0, 1, 0],
            [0, 1, 0],
            [1, 1, 0]
        ],
    ], dtype=np.byte),
    np.array([
        [
            [0, 0, 1],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 0],
            [0, 1, 1]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [1, 0, 0]
        ],
        [
            [1, 1, 0],
            [0, 1, 0],
            [0, 1, 0]
        ]
    ], dtype=np.byte),
    np.array([
        # T Tetrimino
        [
            [0, 1, 0],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 1],
            [0, 1, 0]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [0, 1, 0]
        ],
        [
            [0, 1, 0],
            [1, 1, 0],
            [0, 1, 0]
        ]
    ], dtype=np.byte),
    np.array([
        # Square Tetrimio
        [
            [1, 1],
            [1, 1]
        ],
    ], dtype=np.byte),
    np.array([
        #Ugly L Tetrimio
        [
            [0, 1, 1],
            [1, 1, 0],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 1],
            [0, 0, 1]
        ],
    ], dtype=np.byte),
    np.array([
        [
            [1, 1, 0],
            [0, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 0, 1],
            [0, 1, 1],
            [0, 1, 0]
        ],
    ], dtype=np.byte),
    np.array([
        #Line Tetrimios
        [
            [0, 1, 0, 0],
            [0, 1, 0, 0],
            [0, 1, 0, 0],
            [0, 1, 0, 0]
        ],
        [
            [1, 1, 1, 1],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0]
        ],
    ], dtype=np.byte)
]

# Function to normalize a value between a minimum and maximum value
def normalize(value, min_value, max_value):
    return (value - min_value) / (max_value - min_value)

# Function to calculate the originality of a list of actions
def originality(actions):
    return len(list(set(actions))) / len(actions)

# Function to convert an action to a string
def action_to_str(action):
    actions = ["LEFT", "RIGHT", "ROTATE", "NOTHING"]
    return actions[action]

# Class to represent a game of Tetris
class Tetris:
    # Constructor
    def __init__(self, rows: int, cols: int):
        self.rows = rows
        self.cols = cols
        self.grid = np.zeros((rows, cols), dtype=np.float32)

        # Calculate width of shape to help get the center
        self.piece_pos = ((cols // 2) - 2, 0)
        self.shape_index = random.randint(0, len(TETRIMINOS) - 1)
        self.piece_rotation = 0

        self.is_game_over = False
        self.lines_completed = 0
        self.tick_count = 0

        self.has_rotated = False
        self.moves_repeated = 0

        # List with 5 previous actions
        self.prev_action: int = NOTHING

    # Static method to create a Tetris game from JSON data
    @staticmethod
    def from_json(data):
        rows = len(data["board"])
        cols = len(data["board"][0])

        game = Tetris(rows, cols)
        game.grid = np.array(data["board"], dtype=np.float32)
        game.piece_pos = (data["x"], data["y"])
        game.shape_index = data["current_shape"]
        game.piece_rotation = data["current_rotation"]
        return game

    # Method to convert the game state to a numpy array
    def to_numpy(self):
        normalized_grid = self.grid.flatten()
        normalized_shape_index = np.array([normalize(self.shape_index, 0, len(TETRIMINOS))], dtype=np.float32)
        normalized_piece_rotation = np.array(
            [normalize(self.piece_rotation, 0, len(TETRIMINOS[self.shape_index]))],
            dtype=np.float32
        )
        normalized_piece_pos = np.array([
            normalize(self.piece_pos[0], 0, self.cols),
            normalize(self.piece_pos[1], 0, self.rows)
        ], dtype=np.float32)
        x = np.concatenate((normalized_grid, normalized_shape_index, normalized_piece_rotation, normalized_piece_pos))\
            .reshape((1, -1))\
            .astype('float32')

        return x
    
    # Method to count the number of holes in the grid
    def count_holes(self):
        holes = 0
        # Loop over each column in the grid
        for col in range(self.cols):
            hole_found = False
            # Loop over each row in the grid
            for row in range(self.rows):
                # If a filled block is found, start looking for holes
                if self.grid[row, col] == 1:
                    hole_found = True
                # If a hole is found after a filled block, increment the hole count
                elif hole_found and self.grid[row, col] == 0:
                    holes += 1
        # Return the total number of holes
        return holes

    # Method to calculate the difference in height between adjacent pillars
    def calculate_pillar_height_difference(self):
        # Initialize the heights of each column to 0
        heights = [0] * self.cols
        # Loop over each column in the grid
        for col in range(self.cols):
            # Loop over each row in the grid
            for row in range(self.rows):
                # If a filled block is found, set the height of the column
                if self.grid[row, col] == 1:
                    heights[col] = self.rows - row
                    break
        # Calculate the sum of the differences in height between adjacent pillars
        pillar_difference = sum(abs(heights[i] - heights[i + 1]) for i in range(self.cols - 1))
        # Return the total difference in height
        return pillar_difference

    # Method to calculate the tutorial points
    def tutorial_points(self):
        # Return twice the number of times the piece has been rotated
        return self.has_rotated * 2

    # Method to calculate the score
    def calculate_score(self) -> float:
        # Calculate the score based on the number of lines completed, the number of ticks, the tutorial points, the number of holes, and the difference in pillar height
        return int(
            self.lines_completed*100 + self.tick_count*0.33 + self.tutorial_points() - self.count_holes() - self.calculate_pillar_height_difference()
        )

    # Method to check for a collision with a move
    def collision_with_move(self, pos: tuple[int, int], rotation: int, shape : int):
        # Get the shape of the piece with the given rotation
        current_shape = TETRIMINOS[shape][rotation]

        # Loop over each block in the shape
        for row in range(len(current_shape)):
            for col in range(len(current_shape[row])):
                # If the block is filled
                if (current_shape[row][col] == 1):
                    # Calculate the position of the block on the board
                    boardX = pos[0] + col
                    boardY = pos[1] + row

                    # Check for collisions with the game board boundaries and other Tetriminos
                    if (boardX < 0 or boardX >= self.cols or boardY >= self.rows or
                        (boardY >= 0 and self.grid[boardY][boardX] == 1)):
                        # If a collision is detected, return True
                        return True

        # If no collision is detected, return False
        return False

    # Method to clear completed lines from the grid
    def clear_completed_lines(self):
        # Loop over each row in the grid
        for row in range(len(self.grid)):
            # If all blocks in the row are filled
            if np.all(self.grid[row] == 1):
                # Remove the completed line from the grid
                self.grid = np.delete(self.grid, row, axis=0)

                # Add a new empty line at the top of the grid
                empty_line = np.zeros((1, self.cols), dtype=int)
                self.grid = np.vstack((empty_line, self.grid))

                # Increment the count of completed lines
                self.lines_completed += 1

    # Method to handle a collision with a piece
    def on_piece_collision(self):
        # Save the current piece to the grid
        self.save_piece()
        # Clear any completed lines from the grid
        self.clear_completed_lines()
        # Spawn a new piece
        self.spawn_new()

    # Method to move the current piece down
    def move_down(self):
        # Calculate the new position of the piece
        new_pos = (self.piece_pos[0], self.piece_pos[1]+1)

        # If the piece would not collide at the new position
        if (not self.collision_with_move(new_pos, self.piece_rotation, self.shape_index)):
            # Move the piece to the new position
            self.piece_pos = new_pos
        else:
            # If the piece would collide, handle the collision
            self.on_piece_collision()

    # Method to handle a user input
    def handle_input(self, action: int):
        # Get the current shape of the piece
        shape = TETRIMINOS[self.shape_index]
        # Initialize the new rotation and position to the current values
        new_rotation = self.piece_rotation
        new_pos = self.piece_pos

        # If the action is to move left, decrement the x-coordinate of the position
        if (action == MOVE_LEFT):
            new_pos = (new_pos[0]-1, new_pos[1])
        # If the action is to move right, increment the x-coordinate of the position
        elif (action == MOVE_RIGHT):
            new_pos = (new_pos[0]+1, new_pos[1])
        # If the action is to rotate the piece, increment the rotation
        elif (action == ROTATE):
            new_rotation = (self.piece_rotation + 1) % len(shape)

        # If the piece would not collide at the new position and rotation
        if (not self.collision_with_move(new_pos, new_rotation, self.shape_index)):
            # Move and rotate the piece
            self.piece_pos = new_pos
            self.piece_rotation = new_rotation

    # Method to save the current piece to the grid
    def save_piece(self):
        # Get the current shape of the piece
        shape = TETRIMINOS[self.shape_index][self.piece_rotation]

        # Loop over each block in the shape
        for row in range(len(shape)):
            for col in range(len(shape[row])):
                # Calculate the position of the block in the grid
                x = col + self.piece_pos[0]
                y = row + self.piece_pos[1]

                # If the block is filled
                if shape[row][col] == 1:
                    # Save the block to the grid
                    self.grid[y][x] = 1

    # Method to spawn a new Tetrimino
    def spawn_new(self) -> None:
        # Randomly select a new shape
        new_shape = random.randint(0, len(TETRIMINOS) - 1)
        # Set the initial rotation to 0
        new_rotation = 0
        # Set the initial position to the top center of the grid
        new_pos = ((self.cols // 2) - 2, 0)

        # If the new Tetrimino would collide with existing blocks, end the game
        if (self.collision_with_move(new_pos, new_rotation, new_shape)):
            self.is_game_over = True
        else:
            # Otherwise, set the new Tetrimino as the current piece
            self.shape_index = new_shape
            self.piece_rotation = new_rotation
            self.piece_pos = new_pos

    # Method to advance the game by one tick
    def tick(self, x: int) -> bool:
        # If the game is not over
        if not self.is_game_over:
            # Handle the user's input
            self.handle_input(x)
            # Move the current piece down
            self.move_down()

            # Increment the tick count
            self.tick_count += 1

            # If the user's input was to rotate the piece, set the has_rotated flag
            if x == ROTATE:
                self.has_rotated = True

            # Return True to indicate that the game is still ongoing
            return True
        else:
            # If the game is over, return False
            return False

    # Method to convert the game state to a string
    def to_str(self) -> str:
        # Initialize an empty board
        board: list[list[str]] = [["-"] * self.cols for i in range(self.rows)]
        # Fill in the blocks on the board
        for row in range(self.rows):
            for col in range(self.cols):
                if (self.grid[row, col] == 1):
                    board[row][col] = "#"

        # Draw the current piece on the board
        shape = TETRIMINOS[self.shape_index][self.piece_rotation]
        for row in range(len(shape)):
            for col in range(len(shape[row])):
                x = self.piece_pos[0]
                y = self.piece_pos[1]

                if (shape[row][col] == 1):
                    if ((row+y) < self.rows):
                        board[row+y][col+x] = "1"

        # Convert the board to a string
        text = ""
        for row_list in board:
            text += "|" + "".join(row_list) + "|\n"

        # Return the string representation of the board
        return text

    # Method to define the string representation of the game
    def __repr__(self) -> str:
        # Use the to_str method to convert the game state to a string
        return self.to_str()

# Check if the script is being run directly
if __name__ == "__main__":
    # Initialize a new game of Tetris with 20 rows and 10 columns
    game = Tetris(20, 10)
    # Initialize the previous action to 4 (which might correspond to a specific action in your game)
    prev_x = 4

    # Start the game loop
    while True:
        # Get the user's input
        user_in = input("action: ")

        # If the user input is empty, use the previous action
        if user_in == "":
            x = prev_x
        # If the user input is "s", use the previous action and cycle to the next Tetrimino shape
        elif user_in == "s":
            x = prev_x
            game.shape_index = (game.shape_index + 1) % len(TETRIMINOS)
        # Otherwise, convert the user input to an integer and use it as the action
        else:
            x = int(user_in)

        # Advance the game by one tick with the chosen action
        game.tick(x)
        # Print the current state of the game
        print(game.to_str())
        # Store the chosen action as the previous action for the next loop iteration
        prev_x = x

File: tetris.js
// Define the canvas and context for drawing
const canvas = document.getElementById("tetrisCanvas");
const context = canvas.getContext("2d");

const square_size=20; // Define the size of each square in pixels

const WIDTH2 = 140; // Width of the extra canvas
const HEIGHT2 = 140; // Height of the extra canvas

// Define the size of the game board
const ROWS = 30;
const COLUMNS = 15;

const WIDTH = COLUMNS * square_size; // Width of the game board
const HEIGHT = ROWS * square_size; // Height of the game board

const CELL_SIZE = 30; // Define the size of each cell in pixels

const extraCanvas = document.getElementById("extraCanvas");
extraCanvas.width = WIDTH2;
extraCanvas.height = HEIGHT2;

const ROWS_EXTRA = WIDTH2 / square_size; // Number of rows in the extra canvas
const COLUMNS_EXTRA = HEIGHT2 / square_size; // Number of columns in the extra canvas

const gameBoard = []; // Initialize the game board array

// Fill the game board array with empty cells
for (let row = 0; row < ROWS; row++) {
    const newRow = Array(COLUMNS).fill(0);
    gameBoard.push(newRow);
}

const gameBoardExtra = []; // Initialize the extra game board array

// Fill the extra game board array with empty cells
for (let row = 0; row < ROWS_EXTRA; row++) {
    const newRow = Array(COLUMNS_EXTRA).fill(0);
    gameBoardExtra.push(newRow);
}

const tetriminos = [
    [
        // L Tetrimino
        [
            [1, 0, 0],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 1],
            [0, 1, 0],
            [0, 1, 0]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [0, 0, 1]
        ],
        [
            [0, 1, 0],
            [0, 1, 0],
            [1, 1, 0]
        ],
    ],
    [
        [
            [0, 0, 1],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 0],
            [0, 1, 1]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [1, 0, 0]
        ],
        [
            [1, 1, 0],
            [0, 1, 0],
            [0, 1, 0]
        ]
    ],
    [
        // T Tetrimino
        [
            [0, 1, 0],
            [1, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 1],
            [0, 1, 0]
        ],
        [
            [0, 0, 0],
            [1, 1, 1],
            [0, 1, 0]
        ],

        [
            [0, 1, 0],
            [1, 1, 0],
            [0, 1, 0]
        ]
    ],
    [
        // Square Tetrimio
        [
            [1, 1],
            [1, 1]
        ],
    ],
    [
        //Ugly L Tetrimio
        [
            [0, 1, 1],
            [1, 1, 0],
            [0, 0, 0]
        ],
        [
            [0, 1, 0],
            [0, 1, 1],
            [0, 0, 1]
        ],
    ],
    [
        [
            [1, 1, 0],
            [0, 1, 1],
            [0, 0, 0]
        ],
        [
            [0, 0, 1],
            [0, 1, 1],
            [0, 1, 0]
        ],
    ],
    [    
        //Line Tetrimios
        [
            [0, 1, 0, 0],
            [0, 1, 0, 0],
            [0, 1, 0, 0],
            [0, 1, 0, 0]
        ],
        [
            [1, 1, 1, 1],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0]
        ],
    ]        
];

const tetriminoColors = [
    'blue',    // Color for L Tetrimino
    'red',     // Color for J Tetrimino
    'green',   // Color for T Tetrimino
    'pink',  // Color for Square Tetrimino
    'purple',  // Color for Ugly L Tetrimino
    'orange',  // Color for S Tetrimino
    'cyan'     // Color for Line Tetriminos
]

const nexttetriminoColors = [
    'blue',    // Color for L Tetrimino
    'red',     // Color for J Tetrimino
    'green',   // Color for T Tetrimino
    'pink',  // Color for Square Tetrimino
    'purple',  // Color for Ugly L Tetrimino
    'orange',  // Color for S Tetrimino
    'cyan'     // Color for Line Tetriminos
]

const cellColors = []; // Array to store the color information for each cell on the game board

for (let row = 0; row < ROWS; row++) {
    const newRow = Array(COLUMNS).fill(null);
    cellColors.push(newRow);
}

let ctx;
let next_shape_index = 0;
let next_rotation_index = 0;
let current_shape_index = 0;
let current_rotation_index = 0;
let tetriminoX = 0;
let tetriminoY = 0;
let direction=''
let isMoveRightKeyPressed = false;
let isMoveLeftKeyPressed = false;
let isMoveDownFastKeyPressed = false;
let lastMoveTimestamp = 0;
const moveDelay = 300; // Set the delay in milliseconds
let interval;
let points = 0;
let timer = 0;
let lines_completed = 0;
let refreshrate = 350 ;


// Function to handle keyboard input
function handle_input(event){
    
    if (event.key == ' '){
        rotate();
    }
    if ((event.key == 'd' || event.key === 'D' || event.key === 'ArrowRight') && !isMoveRightKeyPressed) {
        const currentTime = new Date().getTime();
        if (currentTime - lastMoveTimestamp >= moveDelay) {
            isMoveRightKeyPressed = true;
            move_right();
            lastMoveTimestamp = currentTime;
        }
    }
    if ((event.key === 'a' || event.key === 'A' || event.key === 'ArrowLeft') && !isMoveLeftKeyPressed) {
        const currentTime = new Date().getTime();
        if (currentTime - lastMoveTimestamp >= moveDelay) {
            isMoveLeftKeyPressed = true;
            move_left();
            lastMoveTimestamp = currentTime;
        }
    }
    if ((event.key === 's' || event.key === 'S' || event.key === 'ArrowDown') && !isMoveDownFastKeyPressed) {
        isMoveDownFastKeyPressed = true;
        move_down_fast();
    }
    if (event.key == 'r' || event.key == 'R'){
        resetGame()
    }
}

function handle_keyup(event) {
    if (event.key == 'd') {
        isMoveRightKeyPressed = false;
    }
    if (event.key == 'D') {
        isMoveRightKeyPressed = false;
    }
    if (event.key == 'ArrowRight') {
        isMoveRightKeyPressed = false;
    }
    if (event.key == 'a') {
        isMoveLeftKeyPressed = false;
    }
    if (event.key == 'A') {
        isMoveLeftKeyPressed = false;
    }
    if (event.key == 'ArrowLeft') {
        isMoveLeftKeyPressed = false;
    }
    if (event.key == 's') {
        isMoveDownFastKeyPressed = false;
    }
    if (event.key == 'S') {
        isMoveDownFastKeyPressed = false;
    }
    if (event.key == 'ArrowDown') {
        isMoveDownFastKeyPressed = false;
    }

}

function get_ai_move()
{
    let data = JSON.stringify({
        "board": gameBoard,
        "current_shape": current_shape_index,
        "current_rotation": current_rotation_index,
        "x": tetriminoX,
        "y": tetriminoY
    });
    let action = get_move(data);

    switch (action) {
        case "LEFT":
            move_left();
            break;
        case "RIGHT":
            move_right();
            break;
        case "ROTATE":
            rotate();
            break;
        default:
            break;

    }
    console.log(action)
}

function initGame(){
    canvas_element = document.getElementById("tetrisCanvas");
    window.addEventListener("keydown", handle_input)
    window.addEventListener("keyup", handle_keyup);


    // Choose a random Tetrimino type for the initial piece
    current_shape_index = Math.floor(Math.random() * tetriminos.length);

    // Choose a random rotation index for the initial piece
    current_rotation_index = Math.floor(Math.random() * tetriminos[current_shape_index].length);


    // Choose a new Tetrimino type for the next piece
    next_shape_index = Math.floor(Math.random() * tetriminos.length);
    next_rotation_index = Math.floor(Math.random() * tetriminos[next_shape_index].length);


    // Set the initial position at the top of the board
    tetriminoX = Math.floor((COLUMNS - tetriminos[current_shape_index][current_rotation_index][0].length) / 2);
    tetriminoY = 0;

    interval = setInterval(main_loop, refreshrate);

    ctx = canvas_element.getContext("2d");
    ctx.canvas.width=WIDTH; 
    ctx.canvas.height=HEIGHT;
}

function rotate_shape(shape_index, rotation_index) {
    shape_rotations = tetriminos[shape_index];

    return (rotation_index + 1) % shape_rotations.length
}

function rotate() {
    let next_rotation_index = rotate_shape(current_shape_index, current_rotation_index)
    if(!isCollisionWithMove(tetriminoX, tetriminoY, current_shape_index, next_rotation_index)){
        current_rotation_index = next_rotation_index;
    }
}

function render_background(canvas) {
    const ctx = canvas.getContext("2d");
    
    // Clear the canvas
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    
    // Draw the grid lines
    ctx.fillStyle = "white";
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.strokeStyle = "#000000"; // Set the color of the grid lines
    ctx.lineWidth = 1; // Set the width of the grid lines

    // Draw vertical grid lines
    for (let x = 0; x <= canvas.width; x += square_size) {
        ctx.beginPath();
        ctx.moveTo(x, 0);
        ctx.lineTo(x, canvas.height);
        ctx.stroke();
    }

    // Draw horizontal grid lines
    for (let y = 0; y <= canvas.height; y += square_size) {
        ctx.beginPath();
        ctx.moveTo(0, y);
        ctx.lineTo(canvas.width, y);
        ctx.stroke();
    }
}

function next_render_shape(canvas, shape, x, y, colorArray) {
    const ctx = canvas.getContext("2d");
    const color = colorArray[next_shape_index];  // Assign color based on shape index

    ctx.fillStyle = color;
    for (let row = 0; row < shape.length; row++) {
        for (let column = 0; column < shape.length; column++){
            if (shape[row][column] == 1) {
                let canvasX = (column + x) * square_size;
                let canvasY = (row + y) * square_size;
                
                // Draw the filled square
                ctx.fillRect(canvasX, canvasY, square_size, square_size);

                // Draw the grid lines for the square
                ctx.strokeStyle = "#000000";
                ctx.lineWidth = 1;
                ctx.strokeRect(canvasX, canvasY, square_size, square_size);
            }
        }
    }
}

function render_shape(canvas, shape, x, y, colorArray) {
    const ctx = canvas.getContext("2d");
    const color = colorArray[current_shape_index];  // Assign color based on shape index

    ctx.fillStyle = color;
    for (let row = 0; row < shape.length; row++) {
        for (let column = 0; column < shape.length; column++){
            if (shape[row][column] == 1) {
                let canvasX = (column + x) * square_size;
                let canvasY = (row + y) * square_size;
                
                // Draw the filled square
                ctx.fillRect(canvasX, canvasY, square_size, square_size);

                // Draw the grid lines for the square
                ctx.strokeStyle = "#000000";
                ctx.lineWidth = 1;
                ctx.strokeRect(canvasX, canvasY, square_size, square_size);
            }
        }
    }
}

function renderBoard() {
    // Update the color array first
    for (let row = 0; row < gameBoard.length; row++) {
        for (let col = 0; col < gameBoard[row].length; col++) {
            if (gameBoard[row][col] === 1) {
                const colorIndex = gameBoard[row][col] - 1;
                const color = tetriminoColors[colorIndex];
                const isCellOccupied = cellColors[row][col] !== null;

                // Only update the color if the cell is not occupied by another Tetrimino
                if (!isCellOccupied) {
                    cellColors[row][col] = color;
                }
            }
        }
    }

    // Now render the board
    for (let row = 0; row < gameBoard.length; row++) {
        for (let col = 0; col < gameBoard[row].length; col++) {
            if (gameBoard[row][col] === 1) {
                const color = cellColors[row][col];

                ctx.fillStyle = color;
                ctx.fillRect(col * square_size, row * square_size, square_size, square_size);

                // Optionally, you can draw a border around each block
                ctx.strokeStyle = "#000000";
                ctx.lineWidth = 1;
                ctx.strokeRect(col * square_size, row * square_size, square_size, square_size);
            }
        }
    }
}

function stop() {
    // Clear the game board and stop the game loop
    clearInterval(interval);

    // Show the "Game Over" overlay
    document.getElementById("gameOverOverlay").style.display = "flex";
}

function move_down() {
    let newTetriminoY = tetriminoY + 1;

    // Check if the move is valid (no collision and within bounds)
    if (!isCollisionWithMove(tetriminoX, newTetriminoY, current_shape_index, current_rotation_index)) {
        tetriminoY = newTetriminoY;
    } else {
        // If the Tetrimino has hit the bottom, spawn a new one
        spawnNewTetrimino();
    }
}

function move_down_fast() {
    // Move the piece down continuously until a collision is detected
    while (!isCollisionWithMove(tetriminoX, tetriminoY + 1, current_shape_index, current_rotation_index)) {
        tetriminoY++;
    }
}


function move_right() {
    let newTetriminoX = tetriminoX + 1;

    // Check if the move is valid (no collision and within bounds)
    if (!isCollisionWithMove(newTetriminoX, tetriminoY, current_shape_index, current_rotation_index)) {
        tetriminoX = newTetriminoX;

    }
}
function move_left() {
    let newTetriminoX = tetriminoX - 1;

    // Check if the move is valid (no collision and within bounds)
    if (!isCollisionWithMove(newTetriminoX, tetriminoY, current_shape_index, current_rotation_index)) {
        tetriminoX = newTetriminoX;

    }
}

function clearCompletedLines() {
    for (let row = 0; row < ROWS ; row++) {
        if (gameBoard[row].every(cell => cell === 1)) {
            // Remove the completed line
            gameBoard.splice(row, 1);
            
            // Add a new empty line at the top
            gameBoard.unshift(Array(COLUMNS).fill(0));

            // Adjust the cellColors array accordingly
            cellColors.splice(row, 1);
            cellColors.unshift(Array(COLUMNS).fill(null));
            // Increment lines_completed
            lines_completed++;
        }
    }
}

function spawnNewTetrimino() {
    // Save the current Tetrimino on the game board
    saveTetriminoOnBoard();
    
    current_shape_index = next_shape_index;
    current_rotation_index = next_rotation_index;

    // Choose a new Tetrimino type for the next piece
    next_shape_index = Math.floor(Math.random() * tetriminos.length);
    next_rotation_index = Math.floor(Math.random() * tetriminos[next_shape_index].length);

    // Set the initial position at the top of the board
    tetriminoX = Math.floor((COLUMNS - tetriminos[current_shape_index][current_rotation_index][0].length) / 2);
    tetriminoY = 0;

    // Check and clear completed lines
    clearCompletedLines();

    if (isCollisionWithMove(tetriminoX, tetriminoY, current_shape_index, current_rotation_index)) {
         // Show "Game Over" message
        stop(); // Stop the game loop
    }
}

function saveTetriminoOnBoard() {
    const currentShape = tetriminos[current_shape_index][current_rotation_index];

    for (let row = 0; row < currentShape.length; row++) {
        for (let col = 0; col < currentShape[row].length; col++) {
            if (currentShape[row][col] === 1) {
                const boardX = tetriminoX + col;
                const boardY = tetriminoY + row;

                // Save the Tetrimino on the game board
                gameBoard[boardY][boardX] = 1;
                // Update the color array with the Tetrimino's color
                const colorIndex = current_shape_index;
                const color = tetriminoColors[colorIndex];
                cellColors[boardY][boardX] = color;
            }
        }
    }
}

function isCollisionWithMove(newX, newY, shapeIndex, rotationIndex) {
    const currentShape = tetriminos[shapeIndex][rotationIndex];
    for (let row = 0; row < currentShape.length; row++) {
        for (let col = 0; col < currentShape[row].length; col++) {
            if (currentShape[row][col] === 1) {
                const boardX = newX + col;
                const boardY = newY + row;

                // Check for collisions with the game board boundaries and other Tetriminos
                if (
                    boardX < 0 ||
                    boardX >= COLUMNS ||
                    boardY >= ROWS ||
                    (boardY >= 0 && gameBoard[boardY][boardX] === 1)
                ) {
                    return true; // Collision detected
                }
            }
        }
    }

    return false; // No collision detected
}

function clearboard() {
    for (let row = 0; row < ROWS; row++) {
        gameBoard[row].fill(0); // Clear each row by filling with 0
        cellColors[row].fill(null); // Reset cell colors to null
    }
}

function resetGame() {
    // Clear the game board and reset any other necessary variables
    clearboard();

    // Set the initial position at the top of the board
    tetriminoX = Math.floor((COLUMNS - tetriminos[current_shape_index][current_rotation_index][0].length) / 2);
    tetriminoY = 0;

    // Reset game variables
    isMoveRightKeyPressed = false;
    isMoveLeftKeyPressed = false;
    isMoveDownFastKeyPressed = false;
    lastMoveTimestamp = 0;
    timer = 0;
    points = 0;

    // Clear the interval and restart the game loop
    clearInterval(interval);
    interval = setInterval(main_loop, refreshrate);

    // Hide the "Game Over" overlay
    document.getElementById("gameOverOverlay").style.display = "none";
}

function main_loop() {
    ctx.clearRect(0, 0, WIDTH, HEIGHT);
    // Clear the tetrisCanvas and render its background
    const tetrisCanvas = document.getElementById("tetrisCanvas");
    const extraCanvas = document.getElementById("extraCanvas");
    get_ai_move();
    // Move and spawn new shape if necessary
    move_down();

    const currentShape = tetriminos[current_shape_index][current_rotation_index];
    const nextShape = tetriminos[next_shape_index][next_rotation_index];

    render_background(tetrisCanvas);
    // Clear the extraCanvas and render its background
    render_background(extraCanvas);
  
    render_shape(tetrisCanvas, currentShape, tetriminoX, tetriminoY, tetriminoColors);
    // Render the next Tetrimino at a specific location (adjust coordinates as needed)
    next_render_shape(extraCanvas, nextShape, 2, 2, nexttetriminoColors);
  
    renderBoard();

    // Increment timer
    timer += 0.35;

    // Update points based on time survived
    points = Math.floor(timer) + lines_completed * 100;

 
    // Update the display of timer and points
    document.getElementById("pointsValue").textContent = points;
    document.getElementById("timerValue").textContent = Math.floor(timer) + " seconds";
}

window.onload = function(e) {
    initGame()
}

